sound/soc/sof/topology.c

```diff
@@ -1828,122 +1872,45 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
-			scontrol = sbe->dobj.private;
-			break;
-		case SND_SOC_TPLG_TYPE_ENUM:
-			se = (struct soc_enum *)kc->private_value;
-			scontrol = se->dobj.private;
-			break;
-		default:
-			dev_err(sdev->dev, "error: unknown kcontrol type %d in widget %s\n",
-				widget->dobj.widget.kcontrol_type,
-				widget->name);
-			return -EINVAL;
+	if (ipc_data_size) {
+		for (i = 0; i < widget->num_kcontrols; i++) {
+			memcpy(&process->data + offset,
+			       wdata[i].pdata->data,
+			       wdata[i].pdata->size);
+			offset += wdata[i].pdata->size;
 		}
-
-		/* don't include if no private data */
-		pdata = scontrol->control_data->data;
-		if (!pdata)
-			continue;
-
-		/* make sure data is valid - data can be updated at runtime */
-		if (pdata->magic != SOF_ABI_MAGIC)
-			continue;
-
-		memcpy(&process->data + offset, pdata->data, pdata->size);
-		offset += pdata->size;
 	}
 
-process:
 	process->size = ipc_data_size;
 	swidget->private = process;
 
 	ret = sof_ipc_tx_message(sdev->ipc, process->comp.hdr.cmd, process,
 				 ipc_size, r, sizeof(*r));
 
 	if (ret >= 0 && ipc_data_size)
-		return ret;
+		goto err;
 
 	/* send control data with large message supported method */
 	for (i = 0; i < widget->num_kcontrols; i++) {
-		kc = &widget->kcontrol_news[i];
-
-		switch (widget->dobj.widget.kcontrol_type) {
-		case SND_SOC_TPLG_TYPE_MIXER:
-			sm = (struct soc_mixer_control *)kc->private_value;
-			scontrol = sm->dobj.private;
-			break;
-		case SND_SOC_TPLG_TYPE_BYTES:
-			sbe = (struct soc_bytes_ext *)kc->private_value;
-			scontrol = sbe->dobj.private;
-			break;
-		case SND_SOC_TPLG_TYPE_ENUM:
-			se = (struct soc_enum *)kc->private_value;
-			scontrol = se->dobj.private;
-			break;
-		default:
-			dev_err(sdev->dev, "error: unknown kcontrol type %d\n",
-				widget->dobj.widget.kcontrol_type);
-			ret = -EINVAL;
-			goto err;
-		}
-
-		/* don't include if there's no private data */
-		if (!scontrol->control_data->data)
-			continue;
-
-		/* make sure data is valid - data can be updated at runtime */
-		if (scontrol->control_data->data->magic != SOF_ABI_MAGIC)
-			continue;
-
-		/* get data type */
-		switch (scontrol->cmd) {
-		case SOF_CTRL_CMD_VOLUME:
-		case SOF_CTRL_CMD_ENUM:
-		case SOF_CTRL_CMD_SWITCH:
-			ipc_cmd = SOF_IPC_COMP_SET_VALUE;
-			ctrl_type = SOF_CTRL_TYPE_VALUE_CHAN_SET;
-			break;
-		case SOF_CTRL_CMD_BINARY:
-			ipc_cmd = SOF_IPC_COMP_SET_DATA;
-			ctrl_type = SOF_CTRL_TYPE_DATA_SET;
-			break;
-		default:
-			continue;
-		}
-
-		scontrol->readback_offset = 0;
-
-		ret = snd_sof_ipc_set_get_comp_data(sdev->ipc, scontrol,
-						    ipc_cmd, ctrl_type,
-						    scontrol->cmd, true);
-
+		wdata[i].control->readback_offset = 0;
+		ret = snd_sof_ipc_set_get_comp_data(sdev->ipc, wdata[i].control,
+						    wdata[i].ipc_cmd,
+						    wdata[i].ctrl_type,
+						    wdata[i].control->cmd,
+						    true);
 		if (ret != 0) {
 			dev_err(sdev->dev, "error: send control failed\n");
 			goto err;
 		}
 	}
 
```

Comment:
@juimonen there's an error in freeing process here. kfree(process) would be called even if there's no error which will results in swidget->private being NULL and this will cause the component set up fail upon resume.

Reply:
@ranj063 yes good catch, I'll remove the goto's because they make this ugly. I could do this with couple of goto's but that start's to be too much